File: backend/src/main/java/chatogether/ChaTogether/utils/RegexValidator.java
package chatogether.ChaTogether.utils;

import java.util.regex.Pattern;

public class RegexValidator {
    private final static String emailRegex = "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,6}$";
    private final static String usernameRegex = "^[a-zA-Z_][a-zA-Z0-9._]{3,19}$";
    private final static String passwordRegex = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[!?()*\\[\\]+\\-_.,:;<=>@'\"])[A-Za-z\\d!?()*\\[\\]+\\-_.,:;<=>@'\"]{8,}$";
    private final static String nameRegex = "^\\b([A-ZÀ-ÿ][-,a-z. ']+ *)+$";

    public static boolean validateEmail(String email) {
        return Pattern.compile(emailRegex).matcher(email).matches();
    }

    public static boolean validateUsername(String username) {
        return Pattern.compile(usernameRegex).matcher(username).matches();
    }

    public static boolean validatePassword(String password) {
        return Pattern.compile(passwordRegex).matcher(password).matches();
    }

    public static boolean validateName(String name) {
        return Pattern.compile(nameRegex).matcher(name).matches();
    }
}

File: frontend/lib/utils/regex_validator.dart
class RegexValidator {
  final RegExp _usernameRegex = RegExp(r'^[a-zA-Z_][a-zA-Z0-9._]{3,19}$');
  final RegExp _emailRegex =
      RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,6}$');
  final RegExp _passwordRegex = RegExp(
      r'''^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!?()*\[\]+\-_.,:;<=>@'"])[A-Za-z\d!?()*\[\]+\-_.,:;<=>@'"]{8,}$''');
  final RegExp _nameRegex = RegExp(r"^\b([A-ZÀ-ÿ][-,a-z. ']+ *)+$");

  bool validateUsername(String username) {
    return _usernameRegex.hasMatch(username);
  }

  bool validateEmail(String email) {
    return _emailRegex.hasMatch(email);
  }

  bool validatePassword(String password) {
    return _passwordRegex.hasMatch(password);
  }

  bool validateName(String name) {
    return _nameRegex.hasMatch(name);
  }
}
